### PR TITLE
EN-4600 Remote Config Telemetry

### PR DESCRIPTION
**Pull request by pxjohnny**
1) added support for remote config telemetry activity

### CODE REVIEW

File: src/mod_perimeterx.c [hunk 1]
```diff
@@ -345,6 +347,18 @@ static apr_status_t create_health_check(apr_pool_t *p, server_rec *s, px_config
...
+void telemetry_activity_send_init(server_rec *s, px_config *cfg, char *update_reason) {
+    const char *activity_type = ENFORCER_TELEMETRY_ACTIVITY_TYPE;
+    char *activity = config_to_json_string(cfg, update_reason);
+    if (!activity) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, s, "[%s]: telemetry_activity_send_init: create telemetry activity failed", cfg->app_id);
+        return;
+    }
+
+    post_telemetry(cfg->telemetry_api_url, activity, cfg->api_timeout_ms, cfg, s, NULL, NULL);
+    free(activity);
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: where do we allocate this activity?

File: src/mod_perimeterx.c [hunk 2]
```diff
@@ -345,6 +347,18 @@ static apr_status_t create_health_check(apr_pool_t *p, server_rec *s, px_config
     return rv;
 }
 
+void telemetry_activity_send_init(server_rec *s, px_config *cfg, char *update_reason) {
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: we're using this function not only in init, should change the name so it wont confuse

File: src/mod_perimeterx.c [hunk 3]
```diff
@@ -414,27 +428,28 @@ static void *APR_THREAD_FUNC background_remote_config(apr_thread_t *thd, void *d
...
                 conf->app_id = conf->remote_conf->app_id;
-                conf->monitor_mode = !strcmp(conf->remote_conf->module_mode, "blocking") ? false : true; 
+                conf->monitor_mode = !strcmp(conf->remote_conf->module_mode, "blocking") ? false : true;
                 conf->api_timeout_ms = conf->remote_conf->risk_timeout;
                 conf->ip_header_keys = conf->remote_conf->ip_header_keys;
                 conf->sensitive_header_keys = conf->remote_conf->sensitive_header_keys;
                 // ---------------------------------------------------
-                apr_thread_rwlock_unlock(conf->remote_config_rw_mutex);        
+                apr_thread_rwlock_unlock(conf->remote_config_rw_mutex);
+                telemetry_activity_send_init(remote_conf_data->server, conf, "remote_config");
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: change remote_config to be `static const *char` on the top

File: src/mod_perimeterx.c [hunk 4]
```diff
@@ -414,27 +428,28 @@ static void *APR_THREAD_FUNC background_remote_config(apr_thread_t *thd, void *d
...
                 // ---------------------------------------------------
-                // all thread safe operations should be here 
+                // all thread safe operations should be here
                 conf->remote_conf = remote_conf;
                 conf->module_enabled = conf->remote_conf->module_enabled;
                 conf->payload_key = conf->remote_conf->cookie_key;
                 conf->blocking_score = conf->remote_conf->blocking_score;
                 conf->app_id = conf->remote_conf->app_id;
-                conf->monitor_mode = !strcmp(conf->remote_conf->module_mode, "blocking") ? false : true; 
+                conf->monitor_mode = !strcmp(conf->remote_conf->module_mode, "blocking") ? false : true;
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: change remote_config to be static const *char on the top

File: src/mod_perimeterx.c [hunk 5]
```diff
@@ -530,6 +545,9 @@ static apr_status_t px_child_setup(apr_pool_t *p, server_rec *s) {
 
         cfg->curl_pool = curl_pool_create(cfg->pool, cfg->curl_pool_size);
 
+        ap_log_error(APLOG_MARK, APLOG_DEBUG | APLOG_NOERRNO, 0, s, "px_hook_child_init: start init for telemetry_activity_send");
+        telemetry_activity_send_init(vs, cfg, "initial_config");
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: change remote_config to be static const *char on the top

File: src/mod_perimeterx.c [hunk 6]
```diff
@@ -609,6 +627,7 @@ static const char *set_app_id(cmd_parms *cmd, void *config, const char *app_id)
     conf->risk_api_url = apr_pstrcat(cmd->pool, conf->base_url, RISK_API, NULL);
     conf->captcha_api_url = apr_pstrcat(cmd->pool, conf->base_url, CAPTCHA_API, NULL);
     conf->activities_api_url = apr_pstrcat(cmd->pool, conf->base_url, ACTIVITIES_API, NULL);
+    conf->telemetry_api_url = apr_pstrcat(cmd->pool, conf->base_url, TELEMETRY_API, NULL);
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: `TELEMETRY_API` is not configurable, so we can just set it in the default config create function (which is already there)

File: src/px_client.c [hunk 7]
```diff
@@ -26,3 +26,21 @@ CURLcode post_request(const char *url, const char *payload, long timeout, px_con
     ap_log_error(APLOG_MARK, APLOG_ERR, 0, ctx->r->server, "[%s]: post_req_request: post request payload  %s", ctx->app_id, payload);
     return status;
 }
+
+CURLcode post_telemetry(const char *url, const char *payload, long timeout, px_config *conf, server_rec *s, char **response_data, double *request_rtt) {
```
Comment (nitzanpx) [inline src/px_client.c]: we dont need this function, AFAIK we can just use `post_request_helper`, the function `post_request` in general waits for a CURL to be available, we can create these threads during that function or just ask for them and pass

File: src/px_json.c [hunk 8]
```diff
@@ -421,6 +433,83 @@ char *create_json_response(px_config *cfg, request_context *ctx) {
...
+    json_object_set(config_json, "vid_header_name", json_string(cfg->vid_header_name));
+    json_object_set(config_json, "json_response_enabled", json_boolean(cfg->json_response_enabled));
+    json_object_set(config_json, "cors_headers_enabled", json_boolean(cfg->cors_headers_enabled));
+    json_object_set(config_json, "captcha_type", json_integer(cfg->captcha_type));
+    json_object_set(config_json, "monitor_mode", json_boolean(cfg->monitor_mode));
+    json_object_set(config_json, "enable_token_via_header", json_boolean(cfg->enable_token_via_header));
+    json_object_set(config_json, "remote_config_enabled", json_boolean(cfg->remote_config_enabled));
+    json_object_set(config_json, "remote_config_url", json_string(cfg->remote_config_url));
+    json_object_set(config_json, "remote_config_interval_ms", json_integer(cfg->remote_config_interval_ms));
+    json_object_set(config_json, "cookie_key", json_string(cfg->payload_key));
```
Comment (nitzanpx) [inline src/px_json.c]: rather not add this

File: src/px_json.c [hunk 9]
```diff
@@ -421,6 +433,83 @@ char *create_json_response(px_config *cfg, request_context *ctx) {
...
+    json_object_set(config_json, "blocking_score", json_integer(cfg->blocking_score));
+    json_object_set(config_json, "captcha_enabled", json_boolean(cfg->captcha_enabled));
+    json_object_set(config_json, "module_version", json_string(cfg->module_version));
+    json_object_set(config_json, "skip_mod_by_envvar", json_boolean(cfg->skip_mod_by_envvar));
+    json_object_set(config_json, "curl_pool_size", json_integer(cfg->curl_pool_size));
+    json_object_set(config_json, "base_url", json_string(cfg->base_url));
+    json_object_set(config_json, "risk_api_url", json_string(cfg->risk_api_url));
+    json_object_set(config_json, "captcha_api_url", json_string(cfg->captcha_api_url));
+    json_object_set(config_json, "activities_api_url", json_string(cfg->activities_api_url));
+    json_object_set(config_json, "auth_token", json_string(cfg->auth_token));
```
Comment (nitzanpx) [inline src/px_json.c]: rather not sending this one

File: src/mod_perimeterx.c (hunk 6 above)
Comment (pxjohnny) [reply src/mod_perimeterx.c]: its needed because of the use of base_url

**Pull request merged**